etl/src/v2/state/store/postgres.rs

```diff
@@ -0,0 +1,269 @@
+use std::{collections::HashMap, sync::Arc};
+
+use config::shared::{SourceConfig, TlsConfig};
+use postgres::schema::TableId;
+use sqlx::{
+    postgres::{types::Oid as SqlxOid, PgConnectOptions, PgPoolOptions, PgSslMode},
+    prelude::{FromRow, Type},
+    PgPool, Row,
+};
+use thiserror::Error;
+use tokio::sync::RwLock;
+use tokio_postgres::types::PgLsn;
+
+use crate::v2::{
+    pipeline::PipelineId,
+    replication::slot::get_slot_name,
+    state::{
+        store::base::{StateStore, StateStoreError},
+        table::TableReplicationPhase,
+    },
+    workers::base::WorkerType,
+};
+
+#[derive(Debug, Clone, Copy, Type, PartialEq)]
+#[sqlx(type_name = "etl.table_state", rename_all = "snake_case")]
+pub enum TableState {
+    Init,
+    DataSync,
+    FinishedCopy,
+    SyncDone,
+    Ready,
+    Skipped,
+}
+
+#[derive(Debug, Error)]
+pub enum ToTableStateError {
+    #[error("In-memory table replication phase can't be saved in the state store")]
+    InMemoryPhase,
+}
+
+impl TryFrom<TableReplicationPhase> for TableState {
+    type Error = ToTableStateError;
+
+    fn try_from(value: TableReplicationPhase) -> Result<Self, Self::Error> {
+        Ok(match value {
+            TableReplicationPhase::Init => TableState::Init,
+            TableReplicationPhase::DataSync => TableState::DataSync,
+            TableReplicationPhase::FinishedCopy => TableState::FinishedCopy,
+            TableReplicationPhase::SyncDone { .. } => TableState::SyncDone,
+            TableReplicationPhase::Ready => TableState::Ready,
+            TableReplicationPhase::Skipped => TableState::Skipped,
+            _ => return Err(ToTableStateError::InMemoryPhase),
+        })
+    }
+}
+
+#[derive(Debug, FromRow)]
+pub struct ReplicationStateRow {
+    pub pipeline_id: i64,
+    pub table_id: SqlxOid,
+    pub state: TableState,
+}
+
+#[derive(Debug)]
+struct Inner {
+    table_states: HashMap<TableId, TableReplicationPhase>,
+}
+
+#[derive(Debug, Clone)]
+pub struct PostgresStateStore {
+    pipeline_id: PipelineId,
+    source_config: SourceConfig,
+    inner: Arc<RwLock<Inner>>,
+}
+
+impl PostgresStateStore {
+    pub fn new(pipeline_id: PipelineId, source_config: SourceConfig) -> PostgresStateStore {
+        let inner = Inner {
+            table_states: HashMap::new(),
+        };
+        PostgresStateStore {
+            pipeline_id,
+            source_config,
+            inner: Arc::new(RwLock::new(inner)),
+        }
+    }
+
+    async fn connect_to_source(&self) -> Result<PgPool, sqlx::Error> {
+        let SourceConfig {
```

Comment:
If you can, use `PgConnectionConfig` (of the `sqlx` module) and extend it to support also the `PgSslMode::VerifyFull` ssl mode.

Reply:
Great idea. Moved the code there. This also removed some duplication code between `migrations.rs` in replicator and `postgres.rs` state store in etl crate.